File: MusicAssistant/AirTurnInterface/AirTurnExample/AirTurnExample/ViewController.m
//
//  ViewController.m
//  AirTurnExample
//
//  Created by Nick Brook on 11/08/2014.
//  Copyright (c) 2014 AirTurn. All rights reserved.
//

/*
 
 This is the main AirTurn View Controller.
 
 To test background operation, first connect to an airturn, then kill the App using the button provided (to demonstrate the situation when the App is killed by the system when in the background), then press a pedal on the AirTurn.  You will notice music in the iPod starts playing, this has been triggered by this example App.
 
 To support background operation in your own App, simply tick 'Uses Bluetooth LE devices' in the project Capabilities tab.
 
 */

#define AirTurnPlayPauseiPod (1 && !TARGET_IPHONE_SIMULATOR)

#import "ViewController.h"
#import "PortLabel.h"

#import <AirTurnInterface/AirTurnInterface.h>
#import "AirTurnUIConnectionController.h"

#if AirTurnPlayPauseiPod
@import MediaPlayer;
#endif

@interface ViewController () <UIPopoverPresentationControllerDelegate>

@property (weak, nonatomic) IBOutlet UIButton *settingsButton;
@property (strong, nonatomic) IBOutletCollection(PortLabel) NSArray *portLabels;
@property (weak, nonatomic) IBOutlet UITextField *textField;
@property (weak, nonatomic) IBOutlet UIImageView *connectedLED;

- (IBAction)killApp:(id)sender;

- (void)AirTurnEvent:(NSNotification *)notification;
- (void)viewTapped:(UITapGestureRecognizer *)gr;

@end

@implementation ViewController
            
- (void)viewDidLoad {
    [super viewDidLoad];
    
    // To be notified of button events, add an object as an observer of the button event to NSNotificationCenter.
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(AirTurnEvent:) name:AirTurnPedalPressNotification object:nil];
    
    self.connectedLED.highlighted = [AirTurnManager sharedManager].isConnected;
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(connectionStateChanged:) name:AirTurnConnectionStateChangedNotification object:nil];
    
    // add the tap gesture to exit the text field
    UITapGestureRecognizer *gr = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(viewTapped:)];
    gr.numberOfTapsRequired = 1; gr.numberOfTouchesRequired = 1;
    gr.cancelsTouchesInView = NO; gr.delaysTouchesBegan = NO;
    [self.view addGestureRecognizer:gr];
}

- (void)dealloc {
    // remove as an observer
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark - Airturn

- (IBAction)killApp:(id)sender {
    kill(getpid(), SIGKILL);
}

- (void)AirTurnEvent:(NSNotification *)notification {
    AirTurnPort pedal = [notification.userInfo[AirTurnPortNumberKey] intValue];
    if(pedal == AirTurnPortInvalid) return;
    
    PortLabel *curLab = self.portLabels[pedal-1];
    [curLab highlight];
    
#if AirTurnPlayPauseiPod
    if([AirTurnCentral backgroundOperationEnabled]) {
        MPMusicPlayerController *ipod = [MPMusicPlayerController systemMusicPlayer];
        if(ipod.nowPlayingItem == nil) {
            // pick a random song to play
            MPMediaQuery *query = [MPMediaQuery songsQuery];
            [ipod setQueueWithQuery:query];
            MPMediaItem *item = query.items[arc4random_uniform((u_int32_t)query.items.count)];
            ipod.nowPlayingItem = item;
        }
        if(pedal == AirTurnPort1) {
            [ipod play];
        } else if(pedal == AirTurnPort3) {
            [ipod pause];
        }
    }
#endif
}

- (void)connectionStateChanged:(NSNotification *)notification {
    AirTurnConnectionState state = [notification.userInfo[AirTurnConnectionStateKey] intValue];
    self.connectedLED.highlighted = state == AirTurnConnectionStateConnected;
}

- (BOOL)AirTurnUIRequestsDisplay:(AirTurnUIConnectionController *)connectionController {
    [self performSegueWithIdentifier:@"AirTurnUI" sender:self.settingsButton];
    return YES;
}

#pragma mark - Other bits

- (void)viewTapped:(UITapGestureRecognizer *)gr {
    if(!CGRectContainsPoint(self.textField.frame, [gr locationInView:self.view])) {
        [self.textField resignFirstResponder];
    }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // make the AirTurnUIConnectionController handle its presentation in a popover
    UINavigationController *nav = (UINavigationController *)segue.destinationViewController;
    if([nav respondsToSelector:@selector(popoverPresentationController)]) {
        nav.popoverPresentationController.delegate = self;
    }
    [super prepareForSegue:segue sender:sender];
}

- (void)dismiss {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (UIModalPresentationStyle)adaptivePresentationStyleForPresentationController:(UIPresentationController *)controller {
    return UIModalPresentationFullScreen;
}

- (UIViewController *)presentationController:(UIPresentationController *)controller viewControllerForAdaptivePresentationStyle:(UIModalPresentationStyle)style {
    if([controller.presentedViewController isKindOfClass:[UINavigationController class]]) {
        // add a "Done" button to the parent navigation controller
        UIBarButtonItem *bbi = [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"Done", @"AirTurn UI dismiss button in nav controller") style:UIBarButtonItemStyleDone target:self action:@selector(dismiss)];
        UINavigationController *nc = (UINavigationController *)controller.presentedViewController;
        nc.topViewController.navigationItem.leftBarButtonItem = bbi;
    }
    return controller.presentedViewController;
}

-(IBAction)exit:(UIStoryboardSegue *)segue {
}

@end

File: MusicAssistant 12-22-19/AirTurnInterface/AirTurnExample/AirTurnExample-Swift/ViewController.swift
//
//  ViewController.swift
//  AirTurnExample-Swift
//
//  Created by Nick Brook on 18/11/2015.
//  Copyright © 2015 AirTurn. All rights reserved.
//

import UIKit
import AirTurnInterface
import MediaPlayer

#if (arch(i386) || arch(x86_64)) && os(iOS)
let AirTurnPlayPauseiPod = false
#else
let AirTurnPlayPauseiPod = true
#endif

class ViewController: UIViewController, UIPopoverPresentationControllerDelegate, AirTurnUIDelegate {
    
    @IBOutlet weak var settingsButton: UIButton!
    @IBOutlet var portLabels: [PortLabel]!
    @IBOutlet weak var textField: UITextField!
    @IBOutlet weak var connectedLED: UIImageView!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        // To be notified of button events, add an object as an observer of the button event to NSNotificationCenter.
        NotificationCenter.default.addObserver(self, selector: #selector(ViewController.AirTurnEvent(_:)), name: NSNotification.Name.AirTurnPedalPress, object: nil)
        
        self.connectedLED.isHighlighted = AirTurnManager.shared().isConnected
        NotificationCenter.default.addObserver(self, selector: #selector(ViewController.connectionStateChanged(_:)), name: NSNotification.Name.AirTurnConnectionStateChanged, object: nil)
        
        // add the tap gesture to exit the text field
        let gr = UITapGestureRecognizer(target: self, action: #selector(ViewController.viewTapped(_:)))
        gr.numberOfTapsRequired = 1
        gr.numberOfTouchesRequired = 1
        gr.cancelsTouchesInView = false
        gr.delaysTouchesBegan = false
        self.view.addGestureRecognizer(gr)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    // MARK: - AirTurn

    @IBAction func killApp(_ sender: UIButton) {
        kill(getpid(), SIGKILL)
    }
    
    func AirTurnEvent(_ notification: Notification) {
        if let dict = (notification as NSNotification).userInfo as? [String:AnyObject], let portNum = dict[AirTurnPortNumberKey] as? NSNumber, let pedal = AirTurnPort(rawValue: portNum.intValue) {
            if pedal == .portInvalid {
                return
            }
            let curLab = self.portLabels[pedal.rawValue - 1]
            curLab.highlight()
            
            if AirTurnPlayPauseiPod && AirTurnCentral.backgroundOperationEnabled() {
                let ipod = MPMusicPlayerController.systemMusicPlayer()
                if ipod.nowPlayingItem == nil {
                    let query = MPMediaQuery.songs()
                    ipod.setQueue(with: query)
                    if let items = query.items {
                        let item = items[Int(arc4random_uniform(UInt32(items.count)))]
                        ipod.nowPlayingItem = item
                    }
                }
                switch pedal {
                    case .port1: ipod.play()
                    case .port3: ipod.pause()
                    default: break
                }
            }
        }
    }
    
    func connectionStateChanged(_ notification: Notification) {
        if let dict = (notification as NSNotification).userInfo as? [String:AnyObject], let num = dict[AirTurnPortNumberKey] as? NSNumber, let state = AirTurnConnectionState(rawValue: num.intValue) {
            self.connectedLED.isHighlighted = state == .connected
        }
    }
    
    func airTurnUIRequestsDisplay(_ connectionController: AirTurnUIConnectionController) -> Bool {
        self.performSegue(withIdentifier: "AirTurnUI", sender: self.settingsButton)
        return true
    }
    
    // MARK: - Other bits
    
    func viewTapped(_ gr: UITapGestureRecognizer) {
        if !self.textField.frame.contains(gr.location(in: self.view)) {
            self.textField.resignFirstResponder()
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let nav = segue.destination as? UINavigationController {
            nav.popoverPresentationController?.delegate = self
        }
        super.prepare(for: segue, sender: sender)
    }
    
    func dismissSelf() {
        self.dismiss(animated: true, completion: nil)
    }
    
    func adaptivePresentationStyle(for controller: UIPresentationController) -> UIModalPresentationStyle {
        return .fullScreen
    }
    
    func presentationController(_ controller: UIPresentationController, viewControllerForAdaptivePresentationStyle style: UIModalPresentationStyle) -> UIViewController? {
        if let nc = controller.presentedViewController as? UINavigationController {
            let bbi = UIBarButtonItem(title: NSLocalizedString("Done", comment: "AirTurn UI dismiss button in nav controller"), style: .done, target: self, action: #selector(ViewController.dismissSelf))
            nc.topViewController?.navigationItem.leftBarButtonItem = bbi
        }
        return controller.presentedViewController
    }
    
    @IBAction func exit(_ segue: UIStoryboardSegue) {
        
    }
}

